torchao/kv_cache.py

```diff
@@ -0,0 +1,407 @@
+
+        Raises:
+            AssertionError: If seq_idx is not in block_tables or if new_seq_idx is already in block_tables.
+        """
+        if seq_idx not in self.block_tables:
+            raise AssertionError(
+                f"PagedAttentionCache: Sequence index {seq_idx} is not present in the block tables."
+            )
+
+        self.block_tables[new_seq_idx] = copy.deepcopy(self.block_tables[seq_idx])
+        for block_idx in self.block_tables[seq_idx]:
+            self.block_ref_count[block_idx] += 1
+
+    def set_batch2seq_for_prompt_sharing(self, batch_size: int, beam_size: int):
+        """
+        Set the batch2seq mapping for prompt sharing.
+
+        Args:
+            batch_size (int): The batch size.
+            beam_size (int): The beam size.
+        """
+        self.batch2seq = {}
+        for i in range(batch_size):
+            self.batch2seq[i] = [i * beam_size + j for j in range(beam_size)]
+
+    def _reshape_and_cache(
+        self,
+        slot_mapping: torch.Tensor,
+        key_states: torch.Tensor,
+        value_states: torch.Tensor,
+        layer_idx: int,
+    ):
+        """
+        Reshapes and caches the key and value states based on the given slot mapping.
+
+        Args:
+            slot_mapping (List[List[int]]): A list of lists representing the slot mapping.
+            key_states (torch.Tensor): The key states tensor.
+            value_states (torch.Tensor): The value states tensor.
+            layer_idx (int): The index of the layer.
+
+        Returns:
+            None
+        """
+        slot_mapping = slot_mapping.to(torch.int)
+        block_indicies = torch.div(slot_mapping, self.block_size, rounding_mode="floor")
+        block_indicies = block_indicies.cpu().tolist()
+        block_offsets = slot_mapping % self.block_size
+        block_offsets = block_offsets.cpu().tolist()
+        batch_size = key_states.size(0)
+        seq_len = key_states.size(2)
+        for i in range(batch_size):
+            for seq_idx in range(seq_len):
+                block_idx = block_indicies[i][seq_idx]
+                block_offset = block_offsets[i][seq_idx]
+                for head_idx in range(self.num_key_value_heads):
+                    self.key_caches[layer_idx][block_idx, head_idx, block_offset, :] = (
+                        key_states[i, head_idx, seq_idx, :]
+                    )
+                    self.value_caches[layer_idx][
+                        block_idx, head_idx, block_offset, :
+                    ] = value_states[i, head_idx, seq_idx, :]
+
+    def get_seq_length(self, layer_idx: int = 0) -> int:
+        return self.seen_tokens
+
+    def get_max_length(self) -> Optional[int]:
+        """Returns the maximum sequence length of the cached states. PagedAttentionCache does not have a maximum length."""
+        RuntimeError("PagedAttentionCache does not have a maximum sequence length.")
+
+    def update(
```

Comment:
This function will be called inside the model forward. With python implementation with conditionals and loops here, would it work with `torch.compile`?

Reply:
Good question. We need more effort to integrate this PR to the huggingface to validate the end2end functionality for the torch.compile. I suggest we can review this PR in paralel. I will refine it if need.